### PR TITLE
[std.regex] Bit-NFA kickstart engine

### PR DESCRIPTION
**Pull request by DmitryOlshansky**
This adds an alternative "kickstart" engine to locate a prefix of a given pattern.
It is far more general the the original ShiftOr while a being few times slower,
still it's many times faster then any full-blow engine.

UPDATE: it's even ~x2 times faster then grep! Now that's some news :)

### CODE REVIEW

**Comment by DmitryOlshansky** [comment]
screenshot at 2016-05-07 21 37 12

**Comment by DmitryOlshansky** [comment]
Win32 is blocked by https://issues.dlang.org/show_bug.cgi?id=15989

**Comment by DmitryOlshansky** [comment]
Now with immutable regexes!

**Comment by aG0aep6G** [comment]
Looks like you have a typo in the issue number. 9381 isn't about regex.

**Comment by DmitryOlshansky** [comment]
> Looks like you have a typo in the issue number. 9381 isn't about regex.

Right my bad issue is 9391.

**Comment by DmitryOlshansky** [comment]
All green!

File: std/regex/internal/backtracking.d [hunk 1]
```diff
@@ -1282,16 +1317,16 @@ struct CtContext
...
                 code ~= ctSub( `
-                    if (atEnd || !re.charsets[$$].scanFor(front))
+                    if (atEnd || !charsets[$$].scanFor(front))
                         $$
                     $$
                 $$`, ir[0].data, bailOut, addr >= 0 ? "next();" :"", nextInstr);
             break;
         case IR.Trie:
-            if (charsets.length && charsets[ir[0].data].byInterval.length  <= 8)
+            if (charsets.length && charsets[ir[0].data].length  <= 8)
```
Comment (andralex) [inline std/regex/internal/backtracking.d]: This line seems to be not covered.

Also has two spaces before `<=` :)

File: std/regex/internal/bitnfa.d [hunk 2]
```diff
@@ -0,0 +1,732 @@
...
+    static N* locate(N)(uint key, N[] table)
+    {
+        size_t slot = hashOf(key) & (table.length-1);
+        while (table[slot].occupied)
+        {
+            if (table[slot].key == key)
+                break;
+            slot += 1;
+            if (slot == table.length)
+                slot = 0;
```
Comment (andralex) [inline std/regex/internal/bitnfa.d]: uncovered

File: std/regex/internal/bitnfa.d [hunk 3]
```diff
@@ -0,0 +1,732 @@
...
+                break;
+            case GotoEndOr:
+                paths.push(j+IRL!GotoEndOr+ir[j].data);
+                break;
+            case OrEnd, Wordboundary, Notwordboundary, Bof, Bol, Eol, Eof, Nop, GroupStart, GroupEnd:
+                paths.push(j+ir[j].length);
+                break;
+            case LookaheadStart, NeglookaheadStart, LookbehindStart,
+                NeglookbehindStart:
+                paths.push(j + IRL!LookaheadStart + ir[j].data + IRL!LookaheadEnd);
```
Comment (andralex) [inline std/regex/internal/bitnfa.d]: uncovered (this would be important, right?)

File: std/regex/internal/bitnfa.d [hunk 4]
```diff
@@ -0,0 +1,732 @@
...
+                bitMapping[i] = bitCount;
+                break outer;
+            case OrChar:
+                uint s = ir[i].sequence;
+                for (uint j=i; j<i+s; j++)
+                    bitMapping[j] = bitCount;
+                i += (s-1)*IRL!OrChar;
+                bitCount++;
+                if (bitCount == 32)
+                    break outer;
```
Comment (andralex) [inline std/regex/internal/bitnfa.d]: uncovered

File: std/regex/internal/bitnfa.d [hunk 5]
```diff
@@ -0,0 +1,732 @@
...
+                break;
+            case LookaheadStart, NeglookaheadStart, LookbehindStart,
+                NeglookbehindStart:
+                i += IRL!LookaheadStart + ir[i].data;
+                break;
+            case End:
+                finalMask |= 1u<<bitMapping[i];
+                break;
+            case Any:
+                uint mask = 1u<<bitMapping[i];
```
Comment (andralex) [inline std/regex/internal/bitnfa.d]: uncovered

**Comment by andralex** [comment]
@DmitryOlshansky by the new flow you can merge your work now, optionally with follow-up work after the review.

File: std/regex/package.d [hunk 6]
```diff
@@ -320,13 +320,11 @@ public alias StaticRegex(Char) = std.regex.internal.ir.StaticRegex!(Char);
 
     Throws: $(D RegexException) if there were any errors during compilation.
 +/
```
Comment (schuetzm) [inline std/regex/package.d]: This doc comment needs to be moved down to `regex()`; now it refers to the newly inserted `regexPure()`.

**Comment by DmitryOlshansky** [comment]
@andralex Will try to resolve the issues first then merge.

**Comment by DmitryOlshansky** [comment]
Auto-merge toggled on

**Comment by MartinNowak** [comment]
Seems like this caused a regression w/ building Higgs.
[Issue 16626 – [Reg 2.073] recent dmd nightly runs out of memory building Higgs](https://issues.dlang.org/show_bug.cgi?id=16626).

**Comment by DmitryOlshansky** [comment]
> Seems like this caused a regression w/ building Higgs.
> Issue 16626 – [Reg 2.073] recent dmd nightly runs out of memory building Higgs.

The new version is more demanding on memory during CTFE so yeah, most likely it now fails for some patterns. Dunno what to do here would really love to see the new engine sometime soon.

**Comment by MartinNowak** [comment]
Seems to me we should disable that engine in CTFE for now.
Let's continue the discussion in Bugzilla https://issues.dlang.org/show_bug.cgi?id=16626#c2.

**Comment by MartinNowak** [comment]
Seems to have caused another regression.
[Issue 17066 – [REG2.073a] std.regex captures got immutable](https://issues.dlang.org/show_bug.cgi?id=17066)

**Comment by DmitryOlshansky** [comment]
@MartinNowak will look into it, though sounds highly suspicious.

**Comment by DmitryOlshansky** [comment]
@MartinNowak  https://github.com/dlang/phobos/pull/5042

**Pull request merged**